Basic support for GPIO filter banks

commodo opened the pull request:
This patchset adds support for GPIO filter banks via device-tree.
Based on (TX/RX) LO frequency ranges, a different GPIO setting/configuration is selected.
The TX/RX bands can also be selected.

It also supports sequencing of settings, such that a desired sequence of GPIOs may be provided.
The logic is part of the ad9361 driver and can be enabled via Kconfig.
For the SidekiqZ2, it's enabled.

The commits in the series, describe the device-tree bindings in detail.

Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>

commodo commented:
Changelog v1 => v2:
* addressed all checkpatch messages (formatting stuff) https://travis-ci.org/analogdevicesinc/linux/jobs/371556331

commodo commented:
added `WIP` tag;
this PR has some parts split up into PRs #88  and #87  to parallelize some review flow
once this is done, `WIP` tag will be removed

mhennerich commented:
Reinitialize the device causes kernel crash:

# echo 1 > /sys/kernel/debug/iio/iio\:device0/initialize 
Unable to handle kernel NULL pointer dereference at virtual address 00000030
pgd = cbef0000
[00000030] *pgd=0beb6831, *pte=00000000, *ppte=00000000
Internal error: Oops: 17 [#1] PREEMPT SMP ARM
Modules linked in:
CPU: 0 PID: 896 Comm: sh Not tainted 4.9.0-10362-g1b82b0f #237
Hardware name: Xilinx Zynq Platform
task: df681340 task.stack: cbf06000
PC is at ad9361_apply_settings+0x28/0x29c
LR is at ad9361_apply_settings_seq+0x34/0x80
pc : [<c036da54>]    lr : [<c036d9e0>]    psr: a0060013
sp : cbf07ca0  ip : cbdd2320  fp : 10624dd3
r10: c0580294  r9 : df619018  r8 : cbe4f400
r7 : df619018  r6 : cbdd22c0  r5 : 00000008  r4 : df619010
r3 : cbe18118  r2 : df619018  r1 : 00000030  r0 : cbdd22c0
Flags: NzCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment none
Control: 18c5387d  Table: 0bef004a  DAC: 00000051
Process sh (pid: 896, stack limit = 0xcbf06210)
Stack: (0xcbf07ca0 to 0xcbf08000)
7ca0: cbf07d14 cbf07d58 cbe4f400 00000000 c02218f8 cbe18118 00000003 df619010
7cc0: 00000000 df5e77e0 cbdd22c0 cbe4f400 df619018 c0580294 10624dd3 c036d9e0
7ce0: df619010 df5e77b8 cbdd22c0 df619018 00000000 00000002 cbdd3320 c036da80
7d00: 00000000 00000802 00000000 00989680 00000000 cbe18118 cbf07ca0 df619010
7d20: cbdd22c0 cbdd2320 cbe1e210 00000000 00000002 cbdd3320 10624dd3 c036e1b8
7d40: 00000000 00000000 00000000 c036423c ffffffff 00000000 00000002 cbf07dc4
7d60: 00000000 00000002 3a980000 c003ce38 df5e6cdc df5e6cc4 00000000 ffffffff
7d80: cbf07dc4 c003d220 00000000 00000002 cbf07dc4 df5e6cc0 c062dfe0 cbe19600
7da0: 00000002 00000000 0112a880 c003d250 00000000 00000002 cbf07dc4 c01a86bc
7dc0: 00000000 cbe16f80 49040440 00000000 00000000 cbe19600 00000002 49040440
7de0: cbe19f00 c01a8720 cbe19600 00000000 cbe19700 c01a9e48 cbe1e010 cbdd22c0
7e00: 04c4b400 cbe4f400 cbe1e210 c036862c 00000000 cbe16140 cbe1e188 0112a880
7e20: 80060013 cbdd22c0 00000002 cbdd32c0 cbdd22d0 cbdd22d0 c04c8884 00000000
7e40: 000c0fdd c036ba34 cbf07e58 cbf07e5c 00000001 cbdd6f00 00000000 00000001
7e60: 00000a31 cbf07f74 00020241 000081b6 00000041 df089770 00000002 cbf06000
7e80: 00000040 c01bca5c 00000002 00000000 df0a32a0 cbebad90 df0a4ee0 00000000
7ea0: cbf07ea0 cbf07ea0 00000004 00000003 cbf07f74 c036b8a4 cbdd6f00 cbf07f88
7ec0: 00000002 000c0d48 c04c8884 00000000 000c0fdd c0141db0 cbdd6f00 00000000
7ee0: 00000002 c0141d6c cbdd6f00 cbf07f88 000c0d48 c000f084 cbf06000 c00d7278
7f00: 00000000 00000000 00000000 cbf07f10 00002000 c00e7e10 00000000 00000000
7f20: 0000000b cbd06f18 cbd06f00 cbd06f40 df681340 cbf07f60 00000001 cbdd6f00
7f40: cbf07f88 00000002 00000002 00000002 cbdd6f00 cbf07f88 000c0d48 c00d7ecc
7f60: cbdd6f00 000c0d48 00000002 cbdd6f00 cbdd6f00 00000002 000c0d48 c000f084
7f80: cbf06000 c00d8a64 00000000 00000000 00000002 00000000 00000001 000c0d48
7fa0: 00000004 c000eec0 00000000 00000001 00000001 000c0d48 00000002 00000000
7fc0: 00000000 00000001 000c0d48 00000004 000bedac 00000020 000bf294 000c0fdd
7fe0: 00000000 bed7c654 00010e80 b6eeff2c 60060010 00000001 1ff7e861 1ff7ec61
[<c036da54>] (ad9361_apply_settings) from [<c036d9e0>] (ad9361_apply_settings_seq+0x34/0x80)
[<c036d9e0>] (ad9361_apply_settings_seq) from [<c036da80>] (ad9361_apply_settings+0x54/0x29c)
[<c036da80>] (ad9361_apply_settings) from [<c036e1b8>] (ad9361_adjust_tx_ext_band_settings+0x48/0x70)
[<c036e1b8>] (ad9361_adjust_tx_ext_band_settings) from [<c036423c>] (ad9361_tx_rfpll_rate_change+0xa4/0xb0)
[<c036423c>] (ad9361_tx_rfpll_rate_change) from [<c003ce38>] (notifier_call_chain+0x40/0x68)
[<c003ce38>] (notifier_call_chain) from [<c003d220>] (__srcu_notifier_call_chain+0x88/0xa4)
[<c003d220>] (__srcu_notifier_call_chain) from [<c003d250>] (srcu_notifier_call_chain+0x14/0x1c)
[<c003d250>] (srcu_notifier_call_chain) from [<c01a86bc>] (__clk_notify+0x58/0x70)
[<c01a86bc>] (__clk_notify) from [<c01a8720>] (__clk_recalc_rates+0x4c/0x84)
[<c01a8720>] (__clk_recalc_rates) from [<c01a9e48>] (clk_core_set_parent+0x138/0x150)
[<c01a9e48>] (clk_core_set_parent) from [<c036862c>] (ad9361_setup+0x710/0x17bc)
[<c036862c>] (ad9361_setup) from [<c036ba34>] (ad9361_debugfs_write+0x190/0x43c)
[<c036ba34>] (ad9361_debugfs_write) from [<c0141db0>] (full_proxy_write+0x44/0x5c)
[<c0141db0>] (full_proxy_write) from [<c00d7278>] (__vfs_write+0x20/0x108)
[<c00d7278>] (__vfs_write) from [<c00d7ecc>] (vfs_write+0xb8/0x144)
[<c00d7ecc>] (vfs_write) from [<c00d8a64>] (SyS_write+0x40/0x80)
[<c00d8a64>] (SyS_write) from [<c000eec0>] (ret_fast_syscall+0x0/0x3c)
Code: 058d7014 028d7014 e1a06000 158d3014 (e5953028) 
---[ end trace d231ad46caaa6c5b ]---

Welcome to Sidekiq Z2
sidekiqz2 login:

mhennerich commented:
GPIO names are garbage ...

# cat /sys/kernel/debug/gpio 
gpiochip2: GPIOs 890-897, parent: i2c/0-0020, tca6408, can sleep:
 gpio-890 (                    |en_ext_i2c_3v3_n    ) in  hi    
 gpio-891 (                    |reset               ) in  hi    
 gpio-892 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-893 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-894 (                    |lna_sw              ) out lo    
 gpio-895 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-896 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-897 (                    |u�
                                  ���B�w         ) in  lo    

gpiochip1: GPIOs 898-905, parent: i2c/0-0021, tca6408, can sleep:
 gpio-898 (                    |u�
                                  ���B�w         ) in  lo    
 gpio-899 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-900 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-901 (                    |u�
                                  ���B�w         ) in  lo    
 gpio-902 (                    |u�
                                  ���B�w         ) in  lo    
 gpio-903 (                    |u�
                                  ���B�w         ) in  lo    
 gpio-904 (                    |u�
                                  ���B�w         ) in  hi    
 gpio-905 (                    |u�
                                  ���B�w         ) in  hi    

gpiochip0: GPIOs 906-1023, parent: platform/e000a000.gpio, zynq_gpio:
 gpio-921 (                    |?                   ) out lo    
 gpio-958 (                    |ulpi resetb         ) out hi    
 gpio-973 (                    |u�
                                  ���B�w         ) in  hi    
#

commodo commented:
Changelog v2 => v3:
* fixed GPIO names ; apparently `devm_gpio_get()` does not make an internal copy of GPIO names, only a reference; which works fine if assigning constants, but not when passing strings from call-stack
* re-implemented `rx_settings/tx_settings` as linked-lists; previously they were array
   * fixed crash caused via `echo 1 > /sys/kernel/debug/iio/iio\:device0/initialize `  ; the previous logic was faulty, and was accessing an invalid memory region ; this happened because for a short interval the LO TX frequency was 0 Hz and that did not meet any filter ranges
   * this also cleaned up the code a bit
* fixed RX antenna isolation for `tdd_ant_3` ; was always high (value on boot)
* dropped attempt to send GPIO values in groups ; this doesn't work well; GPIOs need their direction set individually first, which defeats the purpose of the `gpiod_set_value_array()` function; any caching mechanism from this point on is just too complicated to consider [long-term] (vs adding a functionality in `gpiolib` to do this)
  * all GPIOs are set individually/sequentially, which means multiple/consecutive I2C writes (sigh)....

thanks to @dbogdan for helping out with some testing

commodo commented:
Changelog v3 => v4:
* rebased on newer master ( 6e5df94f3beb3ecbc6551ca68a49a62673c643b1 ) to include DT updates for SidekiqZ2

mhennerich commented:
Look good to me. One thing can you start documenting the dt bindings here:
linux/Documentation/devicetree/bindings/iio/adc/adi,ad9361.txt? 
You can take the explanations from your commit messages they look good.
Once we have adi,ad9361.txt I'll add the ones from here: 
https://wiki.analog.com/resources/tools-software/linux-drivers/iio-transceiver/ad9361-customization

commodo commented:
Will update

commodo commented:
Update: will rebase and add DT bindings (for filter banks) on top of code from PR #97 once approved + merged

commodo commented:
Changelog v4 => v5:
* rebased on top of newer master (ce60f09383c10c8a07b43307c7dfddfc9ecdb47e )
  * added DT documentation parts for filter banks
* removed WIP tag
